docs/rfcs/rfc-040-relay-first-api.md

```diff
@@ -0,0 +1,1282 @@
+# Relay First API
+
+# Problem
+
+The current API is not consistent, making it harder to learn, understand and use.
+
+For example, the following query mixes Relay spec (`node`, `edge`) with the simple API, as well as cursor and limit based pagination.
+
+```graphql
+query PeopleWithMovies {
+    people(where: { name: "Keanu" }, options: { limit: 10 }) {
+        name
+        moviesConnection(where: { node: { title: "The Matrix" } }, first: 10) {
+            edges {
+                node {
+                    title
+                }
+            }
+        }
+    }
+}
+```
+
+# Solution
+
+To solve this, the GraphQL API will be redesigned to be [Relay](https://relay.dev/graphql/connections.htm) first. Sugar syntax will be added to support simple operations with less verbosity.
+
+We will use the following types for the examples:
+
+```graphql
+# Original types
+type Movie @fulltext(indexes: [{ indexName: "MovieTitle", fields: ["title"] }]) {
+    title: String!
+    alternativeTitles: [String!]
+    released: Int
+    actors: [Person!]! @relationship(type: "ACTED_IN", direction: IN, properties: "ActedIn")
+    director: Person! @relationship(type: "DIRECTED", direction: IN)
+}
+
+type Person {
+    name: String!
+    movies: [Movie!]! @relationship(type: "ACTED_IN", direction: OUT, properties: "ActedIn")
+    directed: Movie @relationship(type: "DIRECTED", direction: OUT)
+}
+
+interface ActedIn @relationshipProperties {
+    year: Int
+}
+```
+
+## Connections API - Query
+
+Query will be done with the following operations:
+
+```graphql
+type Query {
+    moviesConnection(
+        first: Int
+        after: String
+        last: Int
+        before: String
```

Comment:
Is there a reason for having both `first` and `last`? If (correct me if I'm wrong) they are both referring to the nr of elements to return then what happens if the user sets both of them with different values - which one has precedence? I propose to condense them down into a single field, like `take` from the wording commonly used in offset-based pagination.
Also, what happens if the user sets both `after` and `before`?

Reply:
I agree, the cursor connections specification gives freedom of choice about whether to implement `first` and `after` or `last` and `before`... Or both. I think implementing both in the same API is a mistake, and if anything there should be a toggle between the two, bearing in mind that we've never been asked for backwards pagination.

Reply:
This is an example of the API following Relay spec allowing for both navigations. The toggle idea is good, and this does not have to be implemented if not required, I'll make a note in the RFC to clear that up